runtime/src/iree/hal/drivers/cuda/tracing.c

```diff
@@ -169,98 +227,189 @@ void iree_hal_cuda_tracing_context_free(
 void iree_hal_cuda_tracing_context_collect(
     iree_hal_cuda_tracing_context_t* context) {
   if (!context) return;
-  if (context->query_tail == context->query_head) {
-    // No outstanding queries.
+  iree_slim_mutex_lock(&context->event_mutex);
+
+  // No outstanding queries
+  if (!context->submitted_event_list.head) {
+    iree_slim_mutex_unlock(&context->event_mutex);
     return;
   }
   IREE_TRACE_ZONE_BEGIN(z0);
-
-  while (context->query_tail != context->query_head) {
-    // Compute the contiguous range of queries ready to be read.
-    // If the ringbuffer wraps around we'll handle that in the next loop.
-    uint32_t try_query_count =
-        context->query_head < context->query_tail
-            ? context->query_capacity - context->query_tail
-            : context->query_head - context->query_tail;
-    IREE_TRACE_ZONE_APPEND_VALUE_I64(z0, (int64_t)try_query_count);
-
-    // Scan and feed the times to tracy, stopping when we hit the first
-    // unavailable query.
-    uint32_t query_base = context->query_tail;
-    uint32_t read_query_count = 0;
-    for (uint32_t i = 0; i < try_query_count; ++i) {
-      // Ensure the event has completed; will return CUDA_ERROR_NOT_READY if
-      // recorded but not retired or any other deferred error.
-      uint16_t query_id = (uint16_t)(query_base + i);
-      CUevent query_event = context->event_pool[query_id];
-      CUresult result = context->symbols->cuEventQuery(query_event);
+  // submitted_event_list is a list of the head elements for each
+  // command_buffer that has been submitted. Here we loop over
```

Comment:
Nit: "command bufffer that has.." Also reflow this comment a bit to approach almost 80 chars per line?